### DISCUSSION TITLE
libtorch自动下载的包不全

### DISCUSSION THREAD

**Discussion by linshuxi** [xmake-io/xmake#4652] [Q&A]
您好，我发现xmake自动下载的包不全，我用的包是libtorch，如果我用下边的xmake.lua：
add_rules("mode.release")
set_languages("c++17")
add_requires("libtorch")

target("hello")
    add_packages("libtorch")    
    set_kind("binary")
    add_files("src/*.cpp")

最后下载的包目录为：
-a----         2023/11/8     14:24         493568 asmjit.dll
-a----         2023/11/8     14:24         981504 c10.dll
-a----         2023/11/8     14:24        5191168 fbgemm.dll
-a----         2023/11/8     14:29         105984 torch.dll
-a----         2023/11/8     14:28      121593856 torch_cpu.dll
-a----         2023/11/8     14:29         105472 torch_global_deps.dll

但是我从网站https://download.pytorch.org/libtorch/nightly/cpu下载下来的包的目录为：
-a----         2023/10/1     16:02         967680 asmjit.dll
-a----         2023/10/1     16:03        2044416 c10.dll
-a----         2023/10/1     16:02       28091392 fbgemm.dll
-a----         2023/10/1     16:03         521216 fbjni.dll
-a----          2023/3/2     17:26        2047000 libiomp5md.dll
-a----          2023/3/2     17:26          41496 libiompstubs5md.dll
-a----         2023/10/1     16:26         632320 pytorch_jni.dll
-a----         2023/10/1     16:26          25088 torch.dll
-a----         2023/10/1     16:26      706373632 torch_cpu.dll
-a----         2023/10/1     16:02          16384 torch_global_deps.dll
-a----         2020/9/18     19:02         195072 uv.dll

可以看到xmake自动下载的包不全，请问这种情况下这么办呢？
同时，用xmake自动下载的包可以编译，但是运行时总是出错，
error: execv(E:\cpp\hello\build\windows\x64\release\hello.exe ) failed(-1073741502)

请问您能给些建议吗？

**Comment by waruqi** [comment]
可以提 pr 到 xmake-repo 改进 libtorch 包。。包配置里也可以加 path env 绑定环境

**Comment by xq114** [comment]
jni模块没有构建（暂时不支持java），xmake构建时默认用本机的openmp没有链接iomp，libuv单独出去作为依赖了，所以会少几个dll；关于运行时出错可以给一段有问题的源代码吗？

**Reply by linshuxi** [reply]
您好，源码是一个简单的测试：
`#include <iostream>
#include <torch/script.h>
#include <fstream>
#include <stdio.h>
#include <cstdio>


int main(int argc, char** argv)
{
   
   torch::Tensor tensor=torch::ones({2,3});
   torch::jit::script::Module module;

   std::string fn = "E:/xfg/pyprojects/pythonProject2/traced_model.pt";


   std::cout<<tensor<<std::endl;
   return 0;

}`

这段代码如果用自动下载的包，编译会通过，但是运行会提示缺少dll。一种解决方法是从网站https://download.pytorch.org/libtorch/nightly/cpu 上下载libtorch，然后将相应的dll全部拷贝到exe所在的目录。但是这种解决方法有缺点，就是如果用jit导入模型的时候，会提示报错，导入模型的代码是：
`module = torch::jit::load(fn);`

我猜这个错误可能是编译的时候用的libtorch和我从网站https://download.pytorch.org/libtorch/nightly/cpu上下载的libtorch版本不匹配造成的。但是不知道怎样验证这种想法.

**Reply by xq114** [reply]
可以用add_requires("cmake::Torch")来引入自己下载的版本，提前设置环境变量让cmake能找到就行。libtorch自己编译的版本可能还是有点问题，要调一下

**Reply by linshuxi** [reply]
谢谢您哈~！

**Reply by xq114** [reply]
之前libtorch有bug确实无法运行，现在修复了，可以试试xrepo update-repo更新到最新仓库
